<a href="https://colab.research.google.com/github/StevenGarc/Ejercicios_IA/blob/master/Examen_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Parcial Final

1. Seleccionar un dataset de Kaggle para trabajar.

  El dataset seleccionado es 'PREDICTING A PULSAR STAR'

In [ ]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
ObjDt = pd.read_csv("pulsar_stars.csv")
ObjDt.head()

Mean of the integrated profile  ...  target_class
0                       140.562500  ...             0
1                       102.507812  ...             0
2                       103.015625  ...             0
3                       136.750000  ...             0
4                        88.726562  ...             0

[5 rows x 9 columns]

#Descripcion


Cada candidato se describe mediante 8 variables continuas y una variable de clase única. Los primeros cuatro son estadísticas simples obtenidas del perfil de pulso integrado (perfil plegado). Este es un conjunto de variables continuas que describen una versión resuelta en longitud de la señal que se ha promediado tanto en tiempo como en frecuencia. Las cuatro variables restantes se obtienen de manera similar de la curva DM-SNR. Estos se resumen a continuación:

    1 Media del perfil integrado.
    2 Desviación estándar del perfil integrado.
    3 Exceso de curtosis del perfil integrado.
    4 Torcedura del perfil integrado.
    5 Media de la curva DM-SNR.
    6 Desviación estándar de la curva DM-SNR.
    7 Exceso de curtosis de la curva DM-SNR.
    8 Torcedura de la curva DM-SNR.
    9 Clase

# Entrenar y evaluar al sistema usando Árboles de Decisión,Regresión Logística y Redes neuronales

In [ ]:
from sklearn import tree
from sklearn import linear_model
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

In [ ]:
#Balanceo de datos
pos = ObjDt[ObjDt['target_class'] == 1]
neg = ObjDt[ObjDt['target_class'] == 0][:len(pos)]
ObjDt = pd.concat([pos, neg])

In [ ]:
attributes = ObjDt.drop('target_class', axis=1)
labels = ObjDt['target_class']

In [ ]:
att_train, att_test, lab_train, lab_test = train_test_split(attributes, labels, test_size =0.3)

Arbol de decision

In [ ]:
model = tree.DecisionTreeClassifier()
model.fit(att_train,lab_train)
lab_pred= model.predict(att_test)

In [ ]:
print("\nMetrincas Arbol de decision:\n\n"+str(classification_report(lab_test,lab_pred)))


Metrincas Arbol de decision:

              precision    recall  f1-score   support

           0       0.92      0.90      0.91       506
           1       0.89      0.92      0.91       478

    accuracy                           0.91       984
   macro avg       0.91      0.91      0.91       984
weighted avg       0.91      0.91      0.91       984



Regresion Logistica

In [ ]:
model = linear_model.LogisticRegression(solver='newton-cg')
model.fit(att_train,lab_train)
lab_pred= model.predict(att_test)

In [ ]:
print("\nMericas Regresion Logistica:\n\n"+str(classification_report(lab_test,lab_pred)))


Mericas Regresion Logistica:

              precision    recall  f1-score   support

           0       0.91      0.97      0.94       506
           1       0.97      0.90      0.94       478

    accuracy                           0.94       984
   macro avg       0.94      0.94      0.94       984
weighted avg       0.94      0.94      0.94       984



Redes Neuronales

In [ ]:
model = MLPClassifier(random_state=1, max_iter=300)
model.fit(att_train,lab_train)
lab_pred= model.predict(att_test)

In [ ]:
print("\nMetricas Redes Neuronales:\n\n"+str(classification_report(lab_test,lab_pred)))


Metricas Redes Neuronales:

              precision    recall  f1-score   support

           0       0.91      0.93      0.92       506
           1       0.92      0.90      0.91       478

    accuracy                           0.91       984
   macro avg       0.92      0.91      0.91       984
weighted avg       0.91      0.91      0.91       984



# Para el entrenamiento y la evaluación de los modelos utilizar k-fold cross validation

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

In [ ]:
model = tree.DecisionTreeClassifier()
kf = KFold(n_splits=3)
accuracy = cross_val_score(model, att_train,lab_train, cv=kf, scoring="accuracy" )
print('accuracy promedio con K-Fold ', np.average(accuracy))

accuracy promedio con K-Fold  0.9097628580227903


In [ ]:
model = linear_model.LogisticRegression(solver='newton-cg')
kf = KFold(n_splits=3)
accuracy = cross_val_score(model, att_train,lab_train, cv=kf, scoring="accuracy" )
print('accuracy promedio con K-Fold ', np.average(accuracy))

accuracy promedio con K-Fold  0.9459472561566802


In [ ]:
model = MLPClassifier(random_state=1, max_iter=300)
kf = KFold(n_splits=3)
accuracy = cross_val_score(model, att_train,lab_train, cv=kf, scoring="accuracy" )
print('accuracy promedio con K-Fold ', np.average(accuracy))

accuracy promedio con K-Fold  0.9293826780275811


# Analizar con las diferentes métricas:

Basado en el valor accuracy de los diferentes modelos, se puede ver que el más alto es el obtenido por el modelo de regresión logística, esta métrica es relevante puesto que permite saber en qué proporción se clasificaron correctamente los casos de prueba, otra métrica importante para este problema es la precisión que permite identificar qué proporción de las estrellas clasificadas como pulsar en realidad lo son, por ello el mejor de los clasificadores es regresión logística con una precisión del 97% y un accuracy de 94% 